src/core/processing/qgsnativealgorithms.cpp

```diff
@@ -2624,31 +2624,49 @@ QVariantMap QgsRasterLayerUniqueValuesReportAlgorithm::processAlgorithm( const Q
   int band = parameterAsInt( parameters, QStringLiteral( "BAND" ), context );
   QString outputFile = parameterAsFileOutput( parameters, QStringLiteral( "OUTPUT_HTML_FILE" ), context );
 
+  QHash< double, uint > uniqueValues;
+  uint layerWidth = layer->width();
+  uint layerHeight = layer->height();
+  uint layerSize = layerWidth * layerHeight;
 
-  QHash< double, int > uniqueValues;
-  int width = layer->width();
-  int height = layer->height();
-
-  QgsRasterBlock *rasterBlock = layer->dataProvider()->block( band, layer->extent(), width, height );
   int noDataCount = -1;
-  if ( rasterBlock->hasNoDataValue() )
+  if ( layer->dataProvider()->sourceHasNoDataValue( band ) )
     noDataCount = 0;
 
-  for ( int row = 0; row < height; row++ )
+  int maxWidth = 4000;
+  int maxHeight = 4000;
+  int nbBlocksWidth = std::ceil( layerWidth / maxWidth );
+  int nbBlocksHeight = std::ceil( layerHeight / maxHeight );
+  int nbBlocks = nbBlocksWidth * nbBlocksHeight;
+
+  QgsRasterIterator iter( layer->dataProvider()->clone() );
```

Comment:
Better make this thread safe, and move this clone up to the prepareAlgorithm step (store a unique_ptr to it as a private member)